src/lib/seam/connect/seam-paginator.ts

```diff
@@ -0,0 +1,155 @@
+import type { Client } from './client.js'
+import type { SeamHttpRequestOptions } from './options.js'
+import { SeamHttpRequest } from './seam-http-request.js'
+
+interface SeamPaginatorParent {
+  readonly client: Client
+  readonly defaults: Required<SeamHttpRequestOptions>
+}
+
+declare const $brand: unique symbol
+
+type PageCursor = string & { [$brand]: 'SeamPageCursor' }
+
+interface Pagination {
+  readonly hasNextPage: boolean
+  readonly nextPageCursor: PageCursor | null
+  readonly nextPageUrl: string | null
+}
+
+export class SeamPaginator<
+  const TResponse,
+  const TResponseKey extends keyof TResponse,
+> implements AsyncIterable<EnsureReadonlyArray<TResponse[TResponseKey]>>
+{
+  readonly #request: SeamHttpRequest<TResponse, TResponseKey>
+  readonly #parent: SeamPaginatorParent
+
+  constructor(
+    parent: SeamPaginatorParent,
+    request: SeamHttpRequest<TResponse, TResponseKey>,
+  ) {
+    if (request.responseKey == null) {
+      throw new Error(
+        `The ${request.pathname} endpoint does not support pagination`,
+      )
+    }
+    this.#parent = parent
+    this.#request = request
+  }
+
+  async firstPage(): Promise<
+    [EnsureReadonlyArray<TResponse[TResponseKey]>, Pagination]
+  > {
+    return await this.#fetch()
+  }
+
+  async nextPage(
+    nextPageCursor: Pagination['nextPageCursor'],
+  ): Promise<[EnsureReadonlyArray<TResponse[TResponseKey]>, Pagination]> {
+    if (nextPageCursor == null) {
+      throw new Error('Cannot get the next page with a null nextPageCursor')
+    }
+
+    return await this.#fetch(nextPageCursor)
+  }
+
+  readonly #fetch = async (
```

Comment:
I think we could use a private method here:
```suggestion
  async #fetch(
```